integrations/dynatrace/client.py

```diff
@@ -0,0 +1,97 @@
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        logger.info(f"Fetching {resource_key} from {url} with params {params}")
+        try:
+            while True:
+                response = await self.client.get(url, params=params)
+                response.raise_for_status()
+                response_data = response.json()
+                resources = response_data[resource_key]
+                yield resources
+                next_page_key = response_data.get("nextPageKey")
+                if not next_page_key:
+                    break
+                params = {"nextPageKey": next_page_key}
+        except httpx.HTTPStatusError as e:
+            logger.error(
+                f"HTTP error on {url}: {e.response.status_code} - {e.response.text}"
+            )
+            raise
+        except httpx.HTTPError as e:
+            logger.error(f"HTTP error on {url}: {e}")
+            raise
+
+    async def _get_problems(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        async for problems in self._get_paginated_resources(
+            f"{self.host_url}/api/v2/problems", "problems"
+        ):
+            yield problems
+
+    async def _get_slos(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        async for slos in self._get_paginated_resources(
+            f"{self.host_url}/api/v2/slo", "slo"
+        ):
+            yield slos
+
+    async def _get_entity_types(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        async for entity_types in self._get_paginated_resources(
+            f"{self.host_url}/api/v2/entityTypes", "types"
+        ):
+            yield entity_types
+
+    async def _get_entities_from_type(
+        self, type: str
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        async for entities in self._get_paginated_resources(
+            f"{self.host_url}/api/v2/entities",
+            "entities",
+            params={"entitySelector": f'type("{type}")'},
+        ):
+            yield entities
+
+    async def _get_entities(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        async for entity_types in self._get_entity_types():
+            for entity_type in entity_types:
+                async for entities in self._get_entities_from_type(entity_type["type"]):
+                    yield entities
+
+    async def get_resource(
+        self, resource: str, **kwargs: dict[str, Any]
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        RESOURCE_MAPPING = {
+            ResourceKey.PROBLEM.value: self._get_problems,
+            ResourceKey.SLO.value: self._get_slos,
+            ResourceKey.ENTITY.value: self._get_entities,
+        }
+
+        if cache := event.attributes.get(resource):
+            logger.info(f"picking {resource} from cache")
+            yield cache
+            return
+
+        async for resources in RESOURCE_MAPPING[resource]():
+            event.attributes.setdefault(resource, []).extend(resources)
+            yield resources
```

Comment:
This function is over complicating the application. Let's remove it and instead call the individual methods in the main.py

Reply:
I think that method simplifies things better compared to having individual methods. Just like we abstracted away the pagination into a different method, I'm abstracting away the caching instead of duplicating both blocks of code into all the methods.

Reply:
The reason why this function may not be needed is because the application can still run without it. What I see it doing is creating a mapping between the resource keys and the functions that implement each kind. In the event we have say then, this mapping will grow and it wouldn't be sustainable.

Another reason I think we may not need it is on the issue of caching. At the moment, I don't see any kind resync or event that will be triggered twice to benefit from the cached resoures. the only resource that will benefit from caching is the _get_entity_type. But that's also used just once in the _get_entity method

Reply:
What could work even better is finding a way to call the _get_paginated_resources(resource_kind) directly in the main.py. You're free to decide on this

Reply:
So, no need to cache, except for retrieving entity types, yes?